addons/binding/org.openhab.binding.solaredge/src/main/java/org/openhab/binding/solaredge/internal/callback/AbstractCommandCallback.java

```diff
@@ -0,0 +1,182 @@
+     */
+    private StatusUpdateListener listener;
+
+    /**
+     * the constructor
+     *
+     * @param config
+     */
+    public AbstractCommandCallback(SolarEdgeConfiguration config) {
+        this.communicationStatus = new CommunicationStatus();
+        this.config = config;
+    }
+
+    /**
+     * the constructor
+     *
+     * @param config
+     */
+    public AbstractCommandCallback(SolarEdgeConfiguration config, StatusUpdateListener listener) {
+        this(config);
+        this.listener = listener;
+    }
+
+    /**
+     * Log request success
+     */
+    @Override
+    public final void onSuccess(Response response) {
+        super.onSuccess(response);
+        communicationStatus.setHttpCode(HttpStatus.getCode(response.getStatus()));
+        logger.debug("HTTP response {}", response.getStatus());
+    }
+
+    /**
+     * Log request failure
+     */
+    @Override
+    public final void onFailure(Response response, Throwable failure) {
+        super.onFailure(response, failure);
+        try {
+            logger.warn("Request failed: {}", failure.toString());
+            communicationStatus.setHttpCode(Code.INTERNAL_SERVER_ERROR);
+
+            // as we are not allowed to catch Throwables we must only throw Exceptions!
+            if (failure instanceof Exception) {
+                communicationStatus.setError((Exception) failure);
+                throw (Exception) failure;
+            }
+        } catch (SocketTimeoutException | TimeoutException e) {
+            communicationStatus.setHttpCode(Code.REQUEST_TIMEOUT);
+        } catch (UnknownHostException e) {
+            communicationStatus.setHttpCode(Code.BAD_GATEWAY);
+        } catch (Exception e) {
+            communicationStatus.setHttpCode(Code.INTERNAL_SERVER_ERROR);
+        }
+
+    }
+
+    @Override
+    public void onContent(Response response, ByteBuffer content) {
+        super.onContent(response, content);
+        logger.debug("received content, length: {}", getContentAsString().length());
+    }
+
+    @Override
+    public void performAction(HttpClient asyncclient) {
+        Request request = asyncclient.newRequest(getURL()).timeout(config.getAsyncTimeout(), TimeUnit.SECONDS);
+        prepareRequest(request).send(this);
+    }
+
+    /**
+     * @return returns Http Status Code
+     */
+    public CommunicationStatus getCommunicationStatus() {
+        if (communicationStatus.getHttpCode() == null) {
+            communicationStatus.setHttpCode(Code.INTERNAL_SERVER_ERROR);
+        }
+        return communicationStatus;
+    }
+
+    /**
+     * converts the json response into an object structure
+     *
+     * @param jsonInString json string to convert
+     * @param clazz class to map to
+     * @return mapped object
+     */
+    protected final <T extends DataResponse> @Nullable T convertJson(@NonNull String jsonInString, Class<T> clazz) {
+        Gson gson = new Gson();
```

Comment:
Ideally you should reuse the Gson instance on class level, constructing it a relatively heavy operation.

Reply:
fixed it